File: HW 4/handout/code/KDE.m
function [p] = KDE(XTrain,XTest,bw,kernel_type)
    % XTrain: training data, size of nTrain * f
    % XTest: test data, size of nTest * f
    % bw: bandwidth
    % p: estimated density for test data, size nTest * 1
    
    [nTrain,f] = size(XTrain);
    nTest = size(XTest,1);
    p = zeros(nTest,1);
    %% begin 

    %% end
end

File: HW 4/handout/code/KDE.py
import numpy as np
from compute_kernel import *

def KDE(XTrain,XTest,bw,kernel_type):
    """
      XTrain: training data, size of nTrain * f
      XTest: test data, size of nTest * f
      bw: bandwidth
      p: estimated density for test data, size nTest * 1
    """
    K = compute_kernel(XTrain,XTest,bw,kernel_type)
    Ksum = np.sum(K, 0)
        
    return Ksum.reshape(-1, 1)
     
   

